src/infrasys/chronify_time_series_storage.py

```diff
@@ -0,0 +1,346 @@
+        """Construct ChronifyTimeSeriesStorage with an existing database file."""
+        id_manager, uuid_lookup = cls._deserialize_ids(data)
+        store = Store(engine_name=data["engine_name"], file_path=Path(data["filename"]))
+        return cls(store, id_manager, read_only=read_only, uuid_lookup=uuid_lookup)
+
+    @staticmethod
+    def _deserialize_ids(data: dict[str, Any]) -> tuple[IDManager, dict[UUID, int]]:
+        uuid_lookup: dict[UUID, int] = {}
+        max_id = 0
+        for key, val in data["uuid_lookup"].items():
+            uuid_lookup[UUID(key)] = val
+            if val > max_id:
+                max_id = val
+        id_manager = IDManager(next_id=max_id + 1)
+        return id_manager, uuid_lookup
+
+    def get_database_url(self) -> str:
+        """Return the path to the underlying database."""
+        assert self._store.engine.url.database is not None
+        # We don't expect to use an in-memory db.
+        return self._store.engine.url.database
+
+    def get_time_series_directory(self) -> Path:
+        assert self._store.engine.url.database is not None
+        return Path(self._store.engine.url.database).parent
+
+    def add_time_series(
+        self,
+        metadata: TimeSeriesMetadata,
+        time_series: TimeSeriesData,
+        connection: Connection | None = None,
+    ) -> None:
+        if not isinstance(time_series, SingleTimeSeries):
+            msg = f"Bug: need to implement add_time_series for {type(time_series)}"
+            raise NotImplementedError(msg)
+
+        if time_series.uuid in self._uuid_lookup:
+            msg = f"Bug: time series {time_series.uuid} already stored"
+            raise Exception(msg)
+
+        db_id = self._id_manager.get_next_id()
+        df = self._to_dataframe(time_series, db_id)
+        schema = _make_table_schema(time_series, _get_table_name(time_series))
+        # There is no reason to run time checks because we are generating the timestamps
+        # from initial_time, resolution, and length, so they are guaranteed to be correct.
+        self._store.ingest_table(df, schema, connection=connection, skip_time_checks=False)
+        self._uuid_lookup[time_series.uuid] = db_id
+        logger.debug("Added {} to time series storage", time_series.summary)
+
+    def check_timestamps(self, key: TimeSeriesKey, connection: Connection | None = None) -> None:
+        table_name = _get_table_name(key)
+        self._store.check_timestamps(table_name, connection=connection)
+
+    def get_engine_name(self) -> str:
+        """Return the name of the underlying database engine."""
+        return self._store.engine.name
+
+    def get_time_series(
+        self,
+        metadata: TimeSeriesMetadata,
+        start_time: datetime | None = None,
+        length: int | None = None,
+        connection: Connection | None = None,
+    ) -> Any:
+        if isinstance(metadata, SingleTimeSeriesMetadata):
```

Comment:
Would you be open to flip the condition? 

```suggestion
        if not isinstance(metadata, SingleTimeSeriesMetadata):
            raise
        return self._get_single_time_series
```

Reply:
We will soon have several time series types: Deterministic, Probabilistic, Scenarios, NonSequential (from Abodh). I imagine this being a long if/else tree or match/case.

Reply:
sounds good. we can leave this as is for now.